src/main/java/org/opensearch/knn/index/memory/SharedIndexStateManager.java

```diff
@@ -0,0 +1,145 @@
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * Constructor
+     */
+    @VisibleForTesting
+    SharedIndexStateManager() {
+        this.sharedIndexStateCache = new ConcurrentHashMap<>();
+        this.readWriteLock = new ReentrantReadWriteLock();
+    }
+
+    /**
+     * Return a {@link SharedIndexState} associated with the key. If no value exists, it will attempt to create it.
+     * Once returned, the {@link SharedIndexState} will be valid until
+     * {@link SharedIndexStateManager#release(SharedIndexState)} is called. Caller must ensure that this is
+     * called after it is done using it.
+     *
+     * In order to create the shared state, it will use the indexAddress passed in to create the shared state from
+     * using {@link org.opensearch.knn.jni.JNIService#initSharedIndexState(long, KNNEngine)}.
+     *
+     * @param indexAddress Address of index to initialize the shared state from
+     * @param knnEngine engine index belongs to
+     * @return ShareModelContext
+     */
+    public SharedIndexState get(long indexAddress, String modelId, KNNEngine knnEngine) {
+        this.readWriteLock.readLock().lock();
+        // This can be done safely with readLock because the ConcurrentHasMap.computeIfAbsent guarantees:
+        //
+        // "If the specified key is not already associated with a value, attempts to compute its value using the given
+        // mapping function and enters it into this map unless null. The entire method invocation is performed
+        // atomically, so the function is applied at most once per key. Some attempted update operations on this map
+        // by other threads may be blocked while computation is in progress, so the computation should be short and
+        // simple, and must not attempt to update any other mappings of this map."
+        //
+        // Ref:
+        // https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-
+        SharedIndexStateEntry entry = sharedIndexStateCache.computeIfAbsent(modelId, m -> {
+            log.info("Loading entry to shared index state cache for model {}", modelId);
+            long sharedIndexStateAddress = JNIService.initSharedIndexState(indexAddress, knnEngine);
+            return new SharedIndexStateEntry(new SharedIndexState(sharedIndexStateAddress, modelId, knnEngine));
+        });
+        entry.incRef();
+        this.readWriteLock.readLock().unlock();
+        return entry.getSharedIndexState();
+    }
+
+    /**
+     * Indicate that the {@link SharedIndexState} is no longer being used. If nothing else is using it, it will be
+     * removed from the cache and evicted.
+     *
+     * After calling this method, {@link SharedIndexState} should no longer be used by calling thread.
+     *
+     * @param sharedIndexState to return to the system.
+     */
+    public void release(SharedIndexState sharedIndexState) {
```

Comment:
should we wrap this whole function in try and finally to ensure that if any exceptions comes from any of the method we are releasing the write lock?

Reply:
Yes that makes sense